src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs

```diff
@@ -1204,6 +1256,60 @@ async static (state) =>
             return (_success, allGraphs, allRuntimes);
         }
 
+        private static Dictionary<LibraryDependencyIndex, VersionRange>? GetAndIndexPackagesToPrune(LibraryDependencyInterningTable libraryDependencyInterningTable, TargetFrameworkInformation? projectTargetFramework)
+        {
+            Dictionary<LibraryDependencyIndex, VersionRange>? prunedPackageVersions = null;
+
+            if (projectTargetFramework?.PackagesToPrune.Count > 0)
+            {
+                prunedPackageVersions = new Dictionary<LibraryDependencyIndex, VersionRange>(capacity: projectTargetFramework.PackagesToPrune.Count);
+
+                foreach (var item in projectTargetFramework.PackagesToPrune)
+                {
+                    LibraryDependencyIndex depIndex = libraryDependencyInterningTable.Intern(item.Value);
+                    prunedPackageVersions[depIndex] = item.Value.VersionRange;
+                }
+            }
+
+            return prunedPackageVersions;
+        }
+
+        private bool ShouldPrunePackage(IReadOnlyDictionary<LibraryDependencyIndex, VersionRange>? packagesToPrune, FindLibraryEntryResult refItemResult, LibraryDependency dep, LibraryDependencyIndex libraryDependencyIndex, bool isPackage, bool isDirectPackageReferenceFromRootProject)
+        {
+            if (packagesToPrune?.TryGetValue(libraryDependencyIndex, out VersionRange? prunableVersion) == true)
+            {
+                if (dep.LibraryRange!.VersionRange!.Satisfies(prunableVersion!.MaxVersion!))
+                {
+                    if (!isPackage)
+                    {
+                        if (SdkAnalysisLevelMinimums.IsEnabled(
+                            _request.Project!.RestoreMetadata!.SdkAnalysisLevel,
+                            _request.Project.RestoreMetadata.UsingMicrosoftNETSdk,
+                            SdkAnalysisLevelMinimums.PruningWarnings))
+                        {
+                            _logger.Log(RestoreLogMessage.CreateWarning(NuGetLogCode.NU1511, string.Format(CultureInfo.CurrentCulture, Strings.Error_RestorePruningProjectReference, dep.Name)));
+                        }
+                    }
+                    else if (isDirectPackageReferenceFromRootProject)
```

Comment:
I see the design spec says that direct package references of transitive projects will be pruned, but reviewing this PR makes me wonder if this will have unexpected/non-intuitive behaviour for customers?

The BCL rarely has breaking changes, but when there is, customers will now need to reference it in every web/console/gui app, as well as every test project. Rather than just once in the class library project that uses the package.

Reply:
> The BCL rarely has breaking changes, but if I customer really does want to reference an older package, they will now need to reference it in every web/console/gui app, as well as every test project. Rather than just once in the class library project that uses the package.

Even when that's the case, conflict resolution will make latest win.